google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java

```diff
@@ -452,7 +452,13 @@ private static String getAppEngineProjectIdFromMetadataServer() throws IOExcepti
             .setReadTimeout(500)
             .setHeaders(new HttpHeaders().set("Metadata-Flavor", "Google"));
     HttpResponse response = request.execute();
-    return response.parseAsString();
+    return headerContainsMetadataFlavor(response) ? response.parseAsString() : null;
+  }
+
+  @InternalApi("Visible for testing")
+  static boolean headerContainsMetadataFlavor(HttpResponse response) {
+    return response.getHeaders()
+      .getFirstHeaderStringValue("Metadata-Flavor").equals("Google");
```

Comment:
Per [the docs for getFirstHeaderStringValue](https://developers.google.com/api-client-library/java/google-http-java-client/reference/1.20.0/com/google/api/client/http/HttpHeaders#getFirstHeaderStringValue(java.lang.String)), this method returns _null_ when the header isn't found, so this method will either return true (inside GCP) _or_ blow up with a `NullPointerException` elsewhere. (A negative test case, as mentioned below, would catch this bug).